nav2_behavior_tree/plugins/decorator/is_path_longer_on_approach.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2022 Neobotix GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+#include <vector>
+#include "nav2_util/geometry_utils.hpp"
+
+#include "nav2_behavior_tree/plugins/decorator/is_path_longer_on_approach.hpp"
+
+namespace nav2_behavior_tree
+{
+
+IsPathLongerOnApproach::IsPathLongerOnApproach(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf)
+{
+  node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
+}
+
+bool IsPathLongerOnApproach::isPathUpdated(
+  nav_msgs::msg::Path & new_path,
+  nav_msgs::msg::Path & old_path)
+{
+  return new_path != old_path && old_path.poses.size() != 0 &&
+         new_path.poses.size() != 0 &&
+         old_path.poses.back() == new_path.poses.back();
+}
+
+bool IsPathLongerOnApproach::isRobotInGoalProximity(
+  nav_msgs::msg::Path & old_path,
+  double & prox_leng)
+{
+  return nav2_util::geometry_utils::calculate_path_length(old_path, 0) < prox_leng;
+}
+
+bool IsPathLongerOnApproach::isNewPathLonger(
+  nav_msgs::msg::Path & new_path,
+  nav_msgs::msg::Path & old_path,
+  double & length_factor)
+{
+  return nav2_util::geometry_utils::calculate_path_length(new_path, 0) >
+         length_factor * nav2_util::geometry_utils::calculate_path_length(
+    old_path, 0);
+}
+
+inline BT::NodeStatus IsPathLongerOnApproach::tick()
+{
+  getInput("path", new_path_);
+  getInput("prox_leng", prox_leng_);
+  getInput("length_factor", length_factor_);
+
+  // Check if the path is updated and valid, compare the old and the new path length,
+  // given the goal proximity and check if the new path is longer
+  if (isPathUpdated(new_path_, old_path_) && isRobotInGoalProximity(old_path_, prox_leng_) &&
+    isNewPathLonger(new_path_, old_path_, length_factor_) && !first_time_)
+  {
+    const BT::NodeStatus child_state = child_node_->executeTick();
+    switch (child_state) {
+      case BT::NodeStatus::RUNNING:
+        return BT::NodeStatus::RUNNING;
+      case BT::NodeStatus::SUCCESS:
+        first_time_ = true;
+        return BT::NodeStatus::SUCCESS;
+      case BT::NodeStatus::FAILURE:
+        return BT::NodeStatus::FAILURE;
+      default:
+        return BT::NodeStatus::FAILURE;
+    }
+  }
+  first_time_ = false;
+  old_path_ = new_path_;
+  return BT::NodeStatus::SUCCESS;
```

Comment:
https://github.com/ros-planning/navigation2/blob/9b674ff43726b599b327155ec83fe5ba0dd8d50e/nav2_behavior_tree/plugins/decorator/speed_controller.cpp#L108 the other decorators just return `status()` in this situation. 

They also set it to running https://github.com/ros-planning/navigation2/blob/9b674ff43726b599b327155ec83fe5ba0dd8d50e/nav2_behavior_tree/plugins/decorator/speed_controller.cpp#L75 

In general, your implementation is fishy w.r.t. the other decorators, for example the most simple: https://github.com/ros-planning/navigation2/blob/9b674ff43726b599b327155ec83fe5ba0dd8d50e/nav2_behavior_tree/plugins/decorator/rate_controller.cpp

Reply:
I have seen that as well. But this decorator is a bit unique to the rest, this would be obviously success on the condition if the path is longer. That’s the main reason why, I also didn’t insist to change the name, as it is kinda conditional_decorator. I think this is the best possible solution with the current state of the BT API.

Reply:
I have seen that as well. But this decorator is a bit unique to the rest, this would be obviously success on the condition if the path is longer. I think this is the best possible solution with the current state of the BT API.

Reply:
I have seen that as well. But this decorator is a bit unique to the rest, this would be obviously success on the condition if the path is longer. The only place, where running occurs is in L73. That happens when you are waiting. So this particular node, does not have to do the return in that way, I think.